verilog/dut_wrapper.h

```diff
@@ -10,79 +11,78 @@
 using namespace std;
 using namespace sc_core;
 
-enum CallbackPhase {
-  Init,
-  BeforeClk,
-  AfterClk,
-};
-
 template <typename DUT> SC_MODULE(DUTWrapper) {
 public:
-  typedef function<void(DUT *)> Callback;
-  vector<Callback> callback_init;
-  vector<Callback> callback_beforeclk;
-  vector<Callback> callback_afterclk;
+  vector<shared_ptr<Callback>> callbacks;
 
-  void register_callback(CallbackPhase phase, Callback callback) {
-    switch (phase) {
-    case CallbackPhase::Init:
-      this->callback_init.push_back(callback);
-      break;
-    case CallbackPhase::BeforeClk:
-      this->callback_beforeclk.push_back(callback);
-      break;
-    case CallbackPhase::AfterClk:
-      this->callback_afterclk.push_back(callback);
-      break;
-    default:
-      assert(false);
-    }
+  void register_callback(shared_ptr<Callback> callback) {
+    this->callbacks.push_back(callback);
   }
 
   SC_HAS_PROCESS(DUTWrapper);
   DUTWrapper(const sc_module_name &name)
-      : sc_module(name), ctx(new VerilatedContext),
-       dut(new DUT(ctx.get())), tfp(new VerilatedFstC) {
+      : sc_module(name), ctx(new VerilatedContext), dut(new DUT(ctx.get())),
+        tfp(new VerilatedFstC) {
     Verilated::traceEverOn(true);
     ctx->traceEverOn(true);
-    dut->trace(tfp.get(), 99);  // Trace 99 levels of hierarchy (or see below)
+    dut->trace(tfp.get(), 99); // Trace 99 levels of hierarchy (or see below)
     std::string filename = std::string((const char *)name) + "_dump.fst";
     tfp->open(filename.c_str());
+    SC_THREAD(Init);
     SC_THREAD(Executor);
   }
 
- 	~DUTWrapper() {
-		tfp->close();
-	}
+  ~DUTWrapper() { tfp->close(); }
 
   sc_in_clk clk;
+  sc_event event_init;
```

Comment:
No need for this, just Init() at CTOR, and Execute as SC_THREAD

Reply:
Fixed.